From 4377954d0c790c6fb461c689f043d7580086fafb Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Fri, 22 Sep 2017 12:57:08 +0200
Subject: [PATCH 1/1] efi_loader: supply EFI network test

This patch provices an EFI application to check the correct function
of the Simple Network Protocol implementation.

It sends a DHCP request and analyzes the DHCP offer.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 cmd/Kconfig                       | 20 ++++++++++++++++++++
 cmd/bootefi.c                     | 17 +++++++++++++++++
 include/asm-generic/sections.h    |  2 ++
 lib/efi_loader/Makefile           |  7 +++++++
 lib/efi_loader/efi_network_test.c | 23 +++++++++++++++++++++++
 5 files changed, 69 insertions(+)
 create mode 100644 lib/efi_loader/efi_network_test.c

diff --git a/cmd/Kconfig b/cmd/Kconfig
index 3ef9b16b08..e0a925700f 100644
--- a/cmd/Kconfig
+++ b/cmd/Kconfig
@@ -222,6 +222,26 @@ config CMD_BOOTEFI_HELLO
 	  for testing that EFI is working at a basic level, and for bringing
 	  up EFI support on a new architecture.
 
+config CMD_BOOTEFI_NETWORK_TEST_COMPILE
+	bool "Compile an EFI application for network testing"
+	depends on CMD_BOOTEFI && (ARM || X86)
+	default y
+	help
+	  This compiles an EFI application that tests the implemenation of the
+	  Simple Network Protocol (SNP) by sending a DHCP request and reading
+	  the DHCP offer.
+
+	  No additional space will be required in the resulting U-Boot binary
+	  when this option is enabled.
+
+config CMD_BOOTEFI_NETWORK_TEST
+	bool "Allow booting an EFI application for network testing"
+	depends on CMD_BOOTEFI_NETWORK_TEST_COMPILE
+	help
+	  This adds an EFI application that tests the implemenation of the
+	  Simple Network Protocol (SNP) so that it can be used with the
+	  'bootefi nettest' command.
+
 source lib/efi_selftest/Kconfig
 
 config CMD_BOOTMENU
diff --git a/cmd/bootefi.c b/cmd/bootefi.c
index b7087e3da8..05e072e768 100644
--- a/cmd/bootefi.c
+++ b/cmd/bootefi.c
@@ -270,6 +270,19 @@ static int do_bootefi(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 		memcpy((char *)addr, __efi_helloworld_begin, size);
 	} else
 #endif
+#ifdef CONFIG_CMD_BOOTEFI_NETWORK_TEST
+	if (!strcmp(argv[1], "nettest")) {
+		ulong size = __efi_efi_network_test_end -
+			     __efi_efi_network_test_begin;
+
+		saddr = env_get("loadaddr");
+		if (saddr)
+			addr = simple_strtoul(saddr, NULL, 16);
+		else
+			addr = CONFIG_SYS_LOAD_ADDR;
+		memcpy((char *)addr, __efi_efi_network_test_begin, size);
+	} else
+#endif
 #ifdef CONFIG_CMD_BOOTEFI_SELFTEST
 	if (!strcmp(argv[1], "selftest")) {
 		struct efi_loaded_image loaded_image_info = {};
@@ -330,6 +343,10 @@ static char bootefi_help_text[] =
 	"bootefi hello\n"
 	"  - boot a sample Hello World application stored within U-Boot\n"
 #endif
+#ifdef CONFIG_CMD_BOOTEFI_NETWORK_TEST
+	"bootefi nettest\n"
+	"  - boot an EFI network test application stored within U-Boot\n"
+#endif
 #ifdef CONFIG_CMD_BOOTEFI_SELFTEST
 	"bootefi selftest\n"
 	"  - boot an EFI selftest application stored within U-Boot\n"
diff --git a/include/asm-generic/sections.h b/include/asm-generic/sections.h
index b6535705a5..1ee56c52f5 100644
--- a/include/asm-generic/sections.h
+++ b/include/asm-generic/sections.h
@@ -24,6 +24,8 @@ extern char __initdata_begin[], __initdata_end[];
 extern char __start_rodata[], __end_rodata[];
 extern char __efi_helloworld_begin[];
 extern char __efi_helloworld_end[];
+extern char __efi_efi_network_test_begin[];
+extern char __efi_efi_network_test_end[];
 
 /* Start and end of .ctors section - used for constructor calls. */
 extern char __ctors_start[], __ctors_end[];
diff --git a/lib/efi_loader/Makefile b/lib/efi_loader/Makefile
index ddb978f650..00ebf1cb8d 100644
--- a/lib/efi_loader/Makefile
+++ b/lib/efi_loader/Makefile
@@ -9,12 +9,19 @@
 
 CFLAGS_helloworld.o := $(CFLAGS_EFI)
 CFLAGS_REMOVE_helloworld.o := $(CFLAGS_NON_EFI)
+CFLAGS_efi_network_test.o := $(CFLAGS_EFI)
+CFLAGS_efi_network_test.o := $(CFLAGS_NON_EFI)
 
 ifneq ($(CONFIG_CMD_BOOTEFI_HELLO_COMPILE),)
 always += helloworld.efi
 endif
 
+ifneq ($(CONFIG_CMD_BOOTEFI_NETWORK_TEST_COMPILE),)
+always += efi_network_test.efi
+endif
+
 obj-$(CONFIG_CMD_BOOTEFI_HELLO) += helloworld_efi.o
+obj-$(CONFIG_CMD_BOOTEFI_NETWORK_TEST) += efi_network_test_efi.o
 obj-y += efi_image_loader.o efi_boottime.o efi_runtime.o efi_console.o
 obj-y += efi_memory.o efi_device_path_to_text.o efi_device_path.o
 obj-y += efi_file.o efi_variable.o efi_bootmgr.o
diff --git a/lib/efi_loader/efi_network_test.c b/lib/efi_loader/efi_network_test.c
new file mode 100644
index 0000000000..77130a36dd
--- /dev/null
+++ b/lib/efi_loader/efi_network_test.c
@@ -0,0 +1,23 @@
+/*
+ * EFI hello world
+ *
+ * Copyright (c) 2016 Google, Inc
+ * Written by Simon Glass <sjg@chromium.org>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+#include <common.h>
+#include <efi_api.h>
+
+efi_status_t EFIAPI efi_main(efi_handle_t handle,
+			     struct efi_system_table *systable)
+{
+	struct efi_simple_text_output_protocol *con_out = systable->con_out;
+	struct efi_boot_services *boottime = systable->boottime;
+
+	con_out->output_string(con_out, L"Hello, world!\n");
+	boottime->exit(handle, 0, 0, NULL);
+
+	return EFI_SUCCESS;
+}
-- 
2.11.0

